File: frontend/src/engine/capabilities/applyFilter.ts
/**
 * applyFilter – Apply a filter to a table or auto-filter range.
 *
 * Office.js notes:
 * - Filters work on Table objects or via Worksheet.autoFilter.
 * - Column index is 0-based within the table.
 * - Multiple filter types: values, topItems, custom.
 */

import { CapabilityMeta, ApplyFilterParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, resolveSheet } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "applyFilter",
  description: "Apply filters to a table or range",
  mutates: false,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.2",
};

async function handler(
  context: Excel.RequestContext,
  params: ApplyFilterParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { tableNameOrRange, columnIndex, criteria } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would apply filter on column ${columnIndex} of ${tableNameOrRange}`,
    };
  }

  options.onProgress?.("Applying filter...");

  // Probe whether tableNameOrRange is a named Table or a range address.
  // getItemOrNullObject avoids the crash that getItem() would cause.
  const tableObj = context.workbook.tables.getItemOrNullObject(tableNameOrRange);
  tableObj.load("isNullObject");
  await context.sync();

  if (!tableObj.isNullObject) {
    // ── Table path ─────────────────────────────────────────────────────────
    const column = tableObj.columns.getItemAt(columnIndex);
    column.load("name");
    await context.sync();

    if (criteria.filterOn === "values" && criteria.values) {
      column.filter.applyValuesFilter(criteria.values);
    } else if (criteria.filterOn === "custom" && criteria.operator && criteria.value !== undefined) {
      const criterion1 = criteria.operator === "contains"
        ? `*${criteria.value}*`
        : `${mapOperator(criteria.operator)}${criteria.value}`;
      const filterCriteria: Excel.FilterCriteria = {
        filterOn: Excel.FilterOn.custom,
        criterion1,
      };
      column.filter.apply(filterCriteria);
    } else if (criteria.filterOn === "topItems" && criteria.value) {
      column.filter.applyTopItemsFilter(Number(criteria.value));
    }

    await context.sync();

    return {
      stepId: "",
      status: "success",
      message: `Applied ${criteria.filterOn} filter on column ${columnIndex} of table "${tableNameOrRange}"`,
      outputs: { range: tableNameOrRange },
    };
  }

  // ── AutoFilter path (range address) ───────────────────────────────────
  // Use resolveRange/resolveSheet so the correct sheet is used regardless
  // of which sheet is currently active.
  const range = resolveRange(context, tableNameOrRange);
  const sheet = resolveSheet(context, tableNameOrRange);

  if (criteria.filterOn === "values" && criteria.values) {
    sheet.autoFilter.apply(range, columnIndex, {
      filterOn: Excel.FilterOn.values,
      values: criteria.values,
    });
  } else if (criteria.filterOn === "custom" && criteria.operator && criteria.value !== undefined) {
    const criterion1 = criteria.operator === "contains"
      ? `*${criteria.value}*`
      : `${mapOperator(criteria.operator)}${criteria.value}`;
    sheet.autoFilter.apply(range, columnIndex, {
      filterOn: Excel.FilterOn.custom,
      criterion1,
    });
  } else if (criteria.filterOn === "topItems" && criteria.value) {
    sheet.autoFilter.apply(range, columnIndex, {
      filterOn: Excel.FilterOn.topItems,
      criterion1: String(criteria.value),
    });
  }

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Applied autoFilter on ${tableNameOrRange} column ${columnIndex}`,
    outputs: { range: tableNameOrRange },
  };
}

function mapOperator(op: string): string {
  switch (op) {
    case "greaterThan": return ">";
    case "lessThan":    return "<";
    case "equals":      return "=";
    case "contains":    return "*";
    default:            return "=";
  }
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/apply_filter.py
"""applyFilter — set an AutoFilter on a range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    filters = params.get("filters") or []
    if not address:
        return {"status": "error", "message": "applyFilter requires 'range'."}

    rng = resolve_range(ctx.workbook_handle, address)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would apply {len(filters)} filter(s) to {rng.address}."}

    # AutoFilter is set on the Range.api side because xlwings doesn't wrap it.
    sheet_api = rng.sheet.api
    # Clear any existing filter first.
    try:
        if getattr(sheet_api, "AutoFilterMode", False):
            sheet_api.AutoFilterMode = False
    except Exception:
        pass

    if not filters:
        # Toggle filter on only — no criteria.
        rng.api.AutoFilter()
        return {
            "status": "success",
            "message": f"Enabled filter on {rng.address}.",
            "outputs": {"range": rng.address},
        }

    for f in filters:
        col_idx = int(f.get("columnIndex", 0)) + 1  # Excel is 1-based
        criterion = f.get("criterion") or f.get("value")
        operator = f.get("operator", "equals")
        op_map = {"equals": "=", "notEquals": "<>", "greaterThan": ">", "lessThan": "<"}
        op_str = op_map.get(operator, "=")
        criterion_str = f"{op_str}{criterion}" if operator != "equals" else str(criterion)
        try:
            rng.api.AutoFilter(Field=col_idx, Criteria1=criterion_str)
        except Exception as exc:  # noqa: BLE001
            return {"status": "error", "message": f"AutoFilter failed on column {col_idx}: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Applied {len(filters)} filter(s) to {rng.address}.",
        "outputs": {"range": rng.address},
    }


registry.register("applyFilter", handler, mutates=True)
